PORTING.md

```diff
@@ -0,0 +1,327 @@
+
+Create two files in the `xpcc/architecture/platform/driver/clock/sam4` folder:
+
+- `clock.hpp` (see [STM32 version for verbatim](https://github.com/roboterclubaachen/xpcc/blob/develop/src/xpcc/architecture/platform/driver/clock/stm32/clock.hpp.in?ts=4)):
+```cpp
+// your copyright with bsd license
+#include <stdint.h>
+#include "../../../device.hpp"
+#include "../generic/common_clock.hpp"
+
+using namespace xpcc::clock;
+```
+
+- `clock.cpp` ([STM32 version](https://github.com/roboterclubaachen/xpcc/blob/develop/src/xpcc/architecture/platform/driver/clock/stm32/clock.cpp.in?ts=4)) (assuming 8 MHz initial clock):
+```cpp
+// your copyright with bsd license
+#include "clock.hpp"
+
+namespace xpcc
+{
+namespace clock
+{
+
+uint32_t ATTRIBUTE_FASTDATA fcpu(8000000);
+uint32_t ATTRIBUTE_FASTDATA fcpu_kHz(8000);
+uint16_t ATTRIBUTE_FASTDATA fcpu_MHz(8);
+// Cortex-M0: 4000 loops per MHz
+// Cortex-M7: 1000 loops per MHz
+// otherwise: 3000 loops per MHz
+uint16_t ATTRIBUTE_FASTDATA ns_per_loop asm("ns_per_loop") (3000 / 8);
+
+}
+}
+```
+
+### Porting the GPIO driver
+
+The GPIO driver is the most important driver of them all, because it gives you bit-banged versions of SPI and I2C throught the `xpcc::SoftwareSpiMaster` and `xpcc::SoftwareI2cMaster` classes.
+
+The [STM32 GPIO driver is quite complicated](https://github.com/roboterclubaachen/xpcc/blob/develop/src/xpcc/architecture/platform/driver/gpio/stm32/gpio.hpp.in?ts=4), because we're a little weird (lol), but the only methods need to implement right now are [these ones from the GPIO interface](https://github.com/roboterclubaachen/xpcc/blob/develop/src/xpcc/architecture/interface/gpio.hpp?ts=4#L167-L246).
+
+Use the STM32 driver as a template, and modify it. For now you can ignore anything to do with interrupts, analog, alternate functions, `connect` methods, channels, triggers and ports.
+It is recommended that you implement the `configure`, `setOutput`, `setInput` methods for `InputType` and `OutputType`, so that you can put the GPIO into open drain mode for I2C.
+
+The elements from the device file is available as a name-mangled dictionary, with attributes available as their members.
+So the `gpio` elements are places in the `gpio` + `s` = `gpios` dictionary, which can be accessed by iterating over it:
+```jinja2
+%% for gpio in gpios
+    %# Access `port` attribute:
+    %% set port = gpio.port
+    %# Access `id` attribute:
+    %% set pin = gpio.id
+%% endfor
+```
+
+You also need to provide the [`xpcc_gpio_enable()` function here](https://github.com/roboterclubaachen/xpcc/blob/develop/src/xpcc/architecture/platform/driver/gpio/stm32/gpio_enable.cpp.in), that enables the clocks to all GPIOs on this chip.
+```cpp
+extern "C"
+void
+xpcc_gpio_enable(void)
+{
+    // do your magic
+}
+```
+
+Remember, you can always check the generated code for this driver in `xpcc/build/sam4/empty/libxpcc/generated_platform/driver/gpio/sam4/`
```

Comment:
I really like the fact that you are pointing people to the generated files as this is the way I debug my templates and also a serious advantage over preprocessor macros.

However, we have to make sure that people are including the header files that result from the templates, because `scons` will only generate the from the templates if the main target depends on them.